sljit_src/sljitNativeX86_32.c

```diff
@@ -310,26 +310,39 @@ SLJIT_API_FUNC_ATTRIBUTE sljit_s32 sljit_emit_return(struct sljit_compiler *comp
 		SLJIT_SP, 0, SLJIT_SP, 0, SLJIT_IMM, compiler->local_size));
 #endif
 
-	size = 2 + (compiler->scratches > 7 ? (compiler->scratches - 7) : 0) +
+	size = (compiler->scratches > 9 ? (compiler->scratches - 9) : 0) +
 		(compiler->saveds <= 3 ? compiler->saveds : 3);
+	if (size) {
+		inst = (sljit_u8*)ensure_buf(compiler, 1 + size);
+		FAIL_IF(!inst);
+
+		INC_SIZE(size);
+
+		if (compiler->saveds > 0 || compiler->scratches > 11)
+			POP_REG(reg_map[SLJIT_S0]);
+		if (compiler->saveds > 1 || compiler->scratches > 10)
+			POP_REG(reg_map[SLJIT_S1]);
+		if (compiler->saveds > 2 || compiler->scratches > 9)
+			POP_REG(reg_map[SLJIT_S2]);
+	}
+
+	/* Adjust shadow stack if needed.  NB: Since TMP_REG1 is used
+	   to adjust shadow stack, use 4(%esp) to load return address.  */
+	FAIL_IF(adjust_shadow_stack(compiler, SLJIT_UNUSED, 0, 4));
+
+	size = 2;
 #if (defined SLJIT_X86_32_FASTCALL && SLJIT_X86_32_FASTCALL)
 	if (compiler->args > 2)
 		size += 2;
 #else
-	if (compiler->args > 0)
+	/* FIXME: Is this correct?  */
+	if (0 && compiler->args > 0)
 		size += 2;
 #endif
 	inst = (sljit_u8*)ensure_buf(compiler, 1 + size);
 	FAIL_IF(!inst);
-
 	INC_SIZE(size);
 
-	if (compiler->saveds > 0 || compiler->scratches > 11)
-		POP_REG(reg_map[SLJIT_S0]);
-	if (compiler->saveds > 1 || compiler->scratches > 10)
-		POP_REG(reg_map[SLJIT_S1]);
-	if (compiler->saveds > 2 || compiler->scratches > 9)
-		POP_REG(reg_map[SLJIT_S2]);
```

Comment:
Do we need these changes? I thought the assertion below is enough.

Reply:
We have

```
POP some registers    <-- Separate instruction buffer
Adjust shadow stack if needed <-- Separate instruction buffer
POP TMP_REG1
RETURN
```

When allocating the instruction buffer for `POP some registers`, the buffer size must be
precise.  There must be no unused space in  `POP some registers` instruction buffer.
The bug is latent since a single instruction buffer used to covere:

```
POP some registers    <-- Separate instruction buffer
POP TMP_REG1
RETURN
```

with some unused space after RETURN.

Reply:
Ah, I am sorry. I thought this code is related to emit_x86_instruction. So we make `adjust_shadow_stack` mandatory for `sljit_emit_return`? Could we make it an op0 operation, which can be run by the user when it is necessary?

Reply:
adjust_shadow_stack must know exactly where return address is on stack.  How can a user
tell and pass the info to the op0 operation?

Reply:
The op0 has no operand. The point is, the user pass no info, but you can compute everything you need. See sljit_set_context:

https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitNativeX86_32.c#L247

`compiler->local_size` contains the stack frame size except the saved registers and the return address. The following code shows how the stack is adjusted before return:

https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitNativeX86_32.c#L302

So you can load the return address by adding a constant to esp:
mov ebp, [esp + compiler->local_size + saveds_size]

Note: compiler->local_size + saveds_size is a compile time constant.

saveds_size for x86-32:
https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitNativeX86_32.c#L278

When SLJIT_F64_ALIGNMENT flag is set, it is a bit more complicated because you need two loads:
mov ebp, [esp + compiler->local_size]
mov ebp, [ebp + saveds_size]

As for 64 bit, it is easier because you need only one load all the time (formulas are different though).

Reply:
`saveds = (2 + (scratches > 9 ? (scratches - 9) : 0) + (saveds <= 3 ? saveds : 3)) * sizeof(sljit_uw);`

The 2 in this formula represents the return adress and ebp, so you might need to add only 1:
`saveds = (1 + (scratches > 9 ? (scratches - 9) : 0) + (saveds <= 3 ? saveds : 3)) * sizeof(sljit_uw);`

But please check, I just give the general idea, haven't checked it yet.

Reply:
Can you add a testcase with sljit_emit_return with frames skipped?  The current
testcases don't cover this case.

Reply:
I have added a new test.

Reply:
```
static void test68(void)
{
        /* Test skipping returns from fast calls (return type is normal). */
        executable_code code;
        struct sljit_compiler *compiler;
        struct sljit_jump *call, *jump;
        struct sljit_label *label;
        int i;

        if (verbose)
                printf("Run test67\n");
                                   ^^^^^ Should be test68.
````

Reply:
Thanks. I will fix it later.

Reply:
I added SLJIT_SKIP_FRAMES_BEFORE_RETURN.   It passed testes.